### PR TITLE
State commit / cleanup

### PR DESCRIPTION
**Pull request by guozhangwang**
@ymatsuda 

1) Task state commit can happen either when user called context.commit() during the processing of a record in any processor, or when the commit interval has reached. In the first case, the task will commit it state including local state store, consumed offset and produced record; in the latter case, the thread will just commit the states of all tasks it owns.

2) Did not have the threshold on total #. buffered records across tasks, since after chatting to @hachikuji I feel poll(0)'s overhead is very minimal, and asking users to specify the threshold could be hard since there is already a config for per-partition max buffered records.

3) Some minor fixes on metrics recording, state cleanup, etc.

### CODE REVIEW

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java [hunk 1]
```diff
@@ -31,22 +31,24 @@
...
-     * Creates a new instance of RecordQueue
-     *
-     * @param partition        partition
-     * @param source           source node
-     */
     public RecordQueue(TopicPartition partition, SourceNode source) {
         this.partition = partition;
         this.source = source;
+
+        this.fifoQueue = new ArrayDeque<>();
```
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: Why did you move these trivial variable initializations to constructor?
Is it the recommended Kafka coding style? Just wondering.

**Comment by ymatsuda** [comment]
Left a minor comment.
LGTM

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java (hunk 1 above)
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: Not really kafka conventions but just personal coding style.

Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: oh...

Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/RecordQueue.java]: Is there an issue for moving tirivial initialization to constructor?

**Comment by guozhangwang** [comment]
@ymatsuda closing this PR and creating a new one after merging in refactored kstream.
